lib/fsm/utils.go

```diff
@@ -41,9 +41,42 @@ func CanRollback(plan *storage.OperationPlan, phaseID string) error {
 		return trace.BadParameter(
 			"phase %q has already been rolled back", phase.ID)
 	}
+	if !dryRun && !latestRollback(plan.Phases, phaseID) {
+		return trace.BadParameter(
+			"rollback subsequent phases before rolling back phase %q", phase.ID)
+	}
 	return nil
 }
 
+// latestRollback returns true if the phase specified by phaseID is next in line
+// to be rolledback.
+func latestRollback(phases []storage.OperationPhase, phaseID string) bool {
+	if len(phases) == 0 {
+		return false
+	}
+
+	// latest keeps track of the latest phase that has not yet been rolled back.
+	var latest storage.OperationPhase
+	for _, phase := range phases {
+		if latestRollback(phase.Phases, phaseID) {
+			return true
+		}
+		switch {
+		// if we run into an unstarted phase, all later phases should also be unstarted
+		case phase.IsUnstarted():
+			break
```

Comment:
this is equivalent to the empty `case` clause (as it terminates the innermost `switch`):
```go
switch {
case phase.IsUnstarted():
case phase.IsInProgress(), phase.IsFailed(), phase.IsCompleted():
...
}
```
if you needed to terminate the loop - use a label.

Reply:
Oops. Yes, we would want to terminate the loop here.